### ISSUE TITLE
awsS3: missing @fog_connection.endpoint on moduler branch

### ISSUE THREAD

**Issue by yortz** [owen2345/camaleon-cms#210]
I was using the [moduler](https://github.com/owen2345/camaleon-cms/tree/moduler) branch to see if I can add native S3 integration, as previously documented at the end of this issue

https://github.com/owen2345/camaleon-cms/issues/189

I've discovered that for some reason the S3 settings dont' get saved from the "Filesystem Settings" admin pane.

To solve this and enable S3 support I had to manually open the console and define those options and now when I open the admin the "Filesystem Settings" is correctly populated with S3 bucket, region, key, and secret and of course the S3 checkbox is checked.

All good, except for the fact that when I then try to get into the "Media" pane I got welcomed with the following error:

```
NoMethodError in CamaleonCms::Admin::MediaController#index
undefined method `endpoint' for #<Fog::Storage::AWS::Real:0x007f9bf9e495e8>
```

if i debug the @fog_connection everything seem to be setup properly except for the endpoint, so I am asking if is there any place where I can define this or if this is a new bug that I've found, not sure tough if this branch should be considered production ready talking about S3 integration that seems not be working at least for me :( Any support on this would be highly appreciated.

Thanks in advance! 

```
#<Fog::Storage::AWS::Real:70153944582900 @use_iam_profile=nil @instrumentor=nil @instrumentor_name="fog.aws.storage" @connection_options={:debug_response=>true, :headers=>{"User-Agent"=>"fog/1.36.0 fog-core/1.34.0"}, :persistent=>false} @persistent=false @signature_version=4 @path_style=false @region="eu-central-1" @endpoint=nil @host="s3-eu-central-1.amazonaws.com" @scheme="https" @port=443 @aws_access_key_id="AWS_KEY" @aws_session_token=nil @aws_credentials_expire_at=nil @signer=#<Fog::AWS::SignatureV4:0x007f9bf9e59f60 @region="eu-central-1", @service="s3", @aws_access_key_id="AWS_KEY", @hmac=#<Fog::HMAC:0x007f9bf9e59ec0 @key="AWS4Pbr4r5TOcFaGMsva3RuX/pm/Fdo9e4I/eEykshWR", @digest=#<OpenSSL::Digest: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855>, @signer=#<Proc:0x007f9bf9e59e20@/Users/federicocortini/.rvm/gems/ruby-2.2.3/gems/fog-core-1.34.0/lib/fog/core/hmac.rb:28 (lambda)>>> @connection=#<Fog::XML::Connection:0x007f9bf9ec81e0 @excon=#<Excon::Connection:7f9bfbd03a60 @data={:chunk_size=>1048576, :ciphers=>"HIGH:!SSLv2:!aNULL:!eNULL:!3DES", :connect_timeout=>60, :debug_request=>false, :debug_response=>true, :headers=>{"User-Agent"=>"fog/1.36.0 fog-core/1.34.0"}, :idempotent=>false, :instrumentor_name=>"excon", :middlewares=>[Excon::Middleware::ResponseParser, Excon::Middleware::Expects, Excon::Middleware::Idempotent, Excon::Middleware::Instrumentor, Excon::Middleware::Mock], :mock=>false, :nonblock=>true, :omit_default_port=>false, :persistent=>false, :read_timeout=>60, :retry_limit=>4, :ssl_verify_peer=>true, :tcp_nodelay=>false, :thread_safe_sockets=>true, :uri_parser=>URI, :versions=>"excon/0.45.4 (x86_64-darwin14) ruby/2.2.3", :write_timeout=>60, :host=>"BUCKET_NAME.s3-eu-central-1.amazonaws.com", :hostname=>"BUCKET_NAME.s3-eu-central-1.amazonaws.com", :path=>"", :port=>443, :query=>nil, :scheme=>"https"} @socket_key="https://BUCKET_NAME.s3-eu-central-1.amazonaws.com:443">>>
```

**Comment by yortz** [comment]
I confirm this is a bug, the endpoint never get set in the admin I'll see if i can push some quick fix for this, in the meantime for anyone experiencing the same kind of behaviour just edit the endpoint according to your aws region here 

https://github.com/owen2345/camaleon-cms/blob/moduler/app/helpers/camaleon_cms/uploader_helper.rb#L214

also I confirm that the admin pane is totally broken, it's impossible to save aws prefs in the db, so this is another bug.

**Comment by owen2345** [comment]
@yortz 
I am sorry, this branch is still in testing...
I will review this night!
I apology for the inconvenience.

**Comment by yortz** [comment]
no problem, bugs happen :) at this point you can have a look at the following forked [branch](https://github.com/yortz/camaleon-cms/tree/aws) where I've actually fixed the missing endpoint param for S3 and added this required value for S3 to the db configuration options. At this point the only part not working is the save action when clicking on the submit button on the "Fylesystem Settings". I can confirm this is now working both locally and on production. I've opened a pull request in case you want to pull in those fixes.

**Comment by owen2345** [comment]
1) @yortz do you have any error saving the changes/settings (maybe in console)?
2) Please update the gem and try again, this morning after many tests to reproduce your problem I pushed a small change, but I don't know if it solve your problem, because I didn't reproduce the error.

Regards,

**Comment by yortz** [comment]
I can confirm that it saves the options in the db from console, in fact I am just reading s3 settings from a yml file and after that I save those to the db with a rake task.

```
#config/s3.yml
filesystem_s3_access_key: YOUR_KEY
filesystem_s3_secret_key: SECRET
filesystem_region: us-east-1
filesystem_s3_bucket_name: BUCKETNAME
filesystem_type: s3
filesystem_s3_endpoint: https://storagegateway.us-east-1.amazonaws.com
```

and this is the task I am running (this actually needs to be run only once or at least each time you change S3 options values).

```
namespace :s3 do
  desc 'Sets S3 options defined in config/s3.yml'
  task set_options: :environment do
    site = CamaleonCms::Site.first
    options = YAML.load_file('./config/s3.yml')
    options.each {|k,v|
      p "Setting #{k} => #{v}"
      site.set_option "#{k}", v 
    }
  end
end
```

Later I will pull from `moduler` branch and see if the save action is now working on the "Fylesystem Settings" pane in admin.

I will keep you posted ;)

**Comment by yortz** [comment]
Just pulled in latest changes from `moduler` branch I can confirm that the `Submit` button is not working on the "Fylesystem Settings" pane, I've opened a new issue for this, I guess this needs to be debugged via js

**Comment by owen2345** [comment]
Could you please create a repo with your clean project to debug and see the error?

**Comment by yortz** [comment]
sure if you want i can create an empty rails app and add the `aws` branch, but actually I think that you will be able to replicate this same kind of issue on your local machine by switching to the `moduler` branch where you've just pulled in my changes :)
